bundle/config/mutator/configure_preset_defaults.go

```diff
@@ -134,65 +133,7 @@ func findNonUserPath(b *bundle.Bundle) string {
 	return ""
 }
 
-func validateProductionMode(b *bundle.Bundle, isPrincipalUsed bool) diag.Diagnostics {
-	r := b.Config.Resources
-	for i := range r.Pipelines {
-		if r.Pipelines[i].Development {
-			return diag.Errorf("target with 'mode: production' cannot include a pipeline with 'development: true'")
-		}
-	}
-
-	// We need to verify that there is only a single deployment of the current target.
-	// The best way to enforce this is to explicitly set root_path.
-	advice := "set 'workspace.root_path' to make sure only one copy is deployed"
-	adviceDetail := fmt.Sprintf(
-		"A common practice is to use a username or principal name in this path, i.e. use\n"+
-			"\n"+
-			"  root_path: /Workspace/Users/%s/.bundle/${bundle.name}/${bundle.target}",
-		b.Config.Workspace.CurrentUser.UserName,
-	)
-	if !isExplicitRootSet(b) {
-		if isRunAsSet(r) || isPrincipalUsed {
-			// Just setting run_as is not enough to guarantee a single deployment,
-			// and neither is setting a principal.
-			// We only show a warning for these cases since we didn't historically
-			// report an error for them.
-			return diag.Diagnostics{
-				{
-					Severity: diag.Recommendation,
-					Summary:  "target with 'mode: production' should " + advice,
-					Detail:   adviceDetail,
-				},
-			}
-		}
-		return diag.Diagnostics{
-			{
-				Severity: diag.Error,
-				Summary:  "target with 'mode: production' must " + advice,
-				Detail:   adviceDetail,
-			},
-		}
-	}
-	return nil
-}
-
-// Determines whether run_as is explicitly set for all resources.
-// We do this in a best-effort fashion rather than check the top-level
-// 'run_as' field because the latter is not required to be set.
-func isRunAsSet(r config.Resources) bool {
-	for i := range r.Jobs {
-		if r.Jobs[i].RunAs == nil {
-			return false
-		}
-	}
-	return true
-}
-
-func isExplicitRootSet(b *bundle.Bundle) bool {
-	return b.Target != nil && b.Target.Workspace != nil && b.Target.Workspace.RootPath != ""
-}
-
-func (m *processTargetMode) Apply(ctx context.Context, b *bundle.Bundle) diag.Diagnostics {
+func (m *configurePresetDefaults) Apply(ctx context.Context, b *bundle.Bundle) diag.Diagnostics {
 	switch b.Config.Bundle.Mode {
 	case config.Development:
 		diags := validateDevelopmentMode(b)
```

Comment:
Why does not the same reason that moved validateProductionMode out is not applied to validateDevelopmentMode?

Reply:
Well, `transformDevelopmentMode` changes presets.

`validateDevelopentMode` is just like `validateProductionMode` and only does validation. Those seem like they should be grouped together.

Assuming both validations can be grouped under `validate_target_mode`, I would change the name of the current file to something like `apply_target_mode`. The file is really about the target mode functionality; the presets are just a means to an end.

Reply:
👍 We still have to separate into 2 mutators because we need to check bundle configuration (without resources) before we load any resources to give early error messages and change preset settings. In the very end, when all resources are in their final configuration, we need to check that no pipelines are in development mode for production deployment.